### PR TITLE
allow timezone setting on read-only postgres

### LINKED ISSUE #1699: Exception occurs when setting custom timezone on Postgres db w/ read-only connection
Running a custom query `select * from orders;` returns the following:

error

Anyone have any tips on how to fix? I've tried reinstalling, upgrading, restarting the instance and syncing the database.

### PR DESCRIPTION
**Pull request by parisholley**
Fixes #1699

### CODE REVIEW

**Comment by salsakran** [comment]
Hey @parisholley , thanks for the contribution. 

Can you please add some tests that verify this works?

**Comment by camsaul** [comment]
Replacing all occurrences of the `?` in the SQL sounds like it would be a SQL injection issue since we don't verify that the value of the `timezone` setting.

File: src/metabase/driver/generic_sql/query_processor.clj [hunk 1]
```diff
@@ -341,9 +341,10 @@
   "Set the timezone for the current connection."
   [driver settings connection]
   (let [timezone (:report-timezone settings)
-        sql      (sql/set-timezone-sql driver)]
-    (log/debug (u/pprint-to-str 'green [sql timezone]))
-    (jdbc/db-do-prepared connection [sql timezone])))
+        sql      (sql/set-timezone-sql driver)
+        replaced (clojure.string/replace sql "?" (str "'" (re-matches #"[A-z\/_]+" timezone) "'"))]
```
Comment (camsaul) [inline src/metabase/driver/generic_sql/query_processor.clj]: https://github.com/metabase/metabase/wiki/Metabase-Clojure-Style-Guide#require-namespaces-you-use

**Comment by parisholley** [comment]
@camsaul correct, but no good way around this. the regex is a simple attempt to make it injection safe

**Review by camsaul — CHANGES_REQUESTED** [review]
There's a couple things I think need to be addressed here:

1.  I think it's hacky and misleading to have all the various drivers specify a string like

```sql
SET SESSION TIMEZONE TO ?
```

only to have that `?` replaced with a string literal before it gets executed anyway. Someone could be forgiven for thinking the timezone is actually passed as a prepared statement param.

It would make more sense just to make `set-timezone-sql` a plain format string, e.g.

```sql
SET SESSION TIMEZONE TO %s
```

So it's at least obvious `%s` is going to get replaced with a string literal before the statement is executed. Then we can just have a Clojure call to format like

```clojure
(format sql timezone)
```

2. If there is no timezone or the timezone is invalid we shouldn't execute the statement at all, rather than running something like

```sql
SET SESSION TIMEZONE TO ''
```

3.  A regex with a range of `A-z` is definitely not what we want. What we want here is `A-Za-z`. `A-z` means any character point between `A` and `z` is considered valid, which includes all sorts of nasty things like `[`, `\`, `^`, etc.

**Comment by camsaul** [comment]
@parisholley thanks for your contribution. Since we want to get this merged for our next release I went ahead and opened a new PR, #4872, based off of this one with the changes requested. I'm expecting to merge that shortly. Thanks again!

**Pull request closed without merging**